core/templates/domain/exploration/RuleInputTypeFactory.ts

```diff
@@ -0,0 +1,158 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Factory to determine correct type of RuleInputs
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { IRuleInput } from 'domain/exploration/RuleObjectFactory';
+import { INumberWithUnitsBackendDict } from
+  'domain/objects/NumberWithUnitsObjectFactory';
+import { IFractionDict } from
+  'domain/objects/FractionObjectFactory.ts';
+/* eslint-disable max-len */
+import { INote } from 'extensions/interactions/MusicNotesInput/directives/music-notes-input-rules.service';
+/* eslint-enable max-len */
+import { IGraphBackendDict } from
+  'extensions/interactions/GraphInput/directives/graph-detail.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class RuleInputTypeFactory {
+  _isGraph(variable: Object): variable is IGraphBackendDict {
+    if (this._isString(variable) || this._isNumber(variable)) {
+      return false;
+    }
+
+    return 'vertices' in variable;
+  }
+
+  _isNoteArray(variable: Object): variable is INote[] {
+    if (Array.isArray(variable)) {
+      if (variable.length === 0) {
+        return true;
+      }
+
+      if (typeof variable[0] === 'string' || typeof variable[0] === 'number') {
```

Comment:
There are other datatypes as well, right? That can come here.
Can't you do something like `if typeof variable[0] !== <dataType>`? If not, I think this is fine.

Also, do the same check for all variables in the array, currently it is done only on the first element?

Reply:
@aks681 Actually we don't have to check everything completely because we know `variable` will be of type `IRuleInput` which is `(string | number | IFractionDict |
  INumberWithUnitsBackendDict | string[] | INote[] |
  number[] | IGraphBackendDict| string[][])`. So we just have to differentiate among these types and checking the first element would do that, right?

Reply:
Ditto - add comments.

Reply:
@nishantwrp As @ankita240796 said, add comments for all these. But, are we validating anywhere that `IRuleInput` will just be those and not anything else (apart from just mentioning its types)?